### ISSUE TITLE
Fallback does not work when SERVFAIL occurs

### ISSUE THREAD

**Issue by agneevX** [AdguardTeam/dnsproxy#257]
I have dnsproxy configured with a fallback option, but when the primary address returns SERVFAIL (often due to a timeout), it does not fallback to the configured server.

**Comment by ameshkov** [comment]
Yep, the fallback currently is only used when a network error occurs.

I am not sure about changing that logic globally, or maybe we should add a config setting that controls when to fallback.

**Comment by agneevX** [comment]
> Yep, the fallback currently is only used when a network error occurs.

Can you please elaborate on what that means, and what conditions surround it?

**Comment by ameshkov** [comment]
It means any network error (a timeout, connection reset, etc), i.e. it means that the DNS request was not successful.

And SERVFAIL means that the request was successful, it's just the server responded with a message with SERVFAIL status.

**Comment by agneevX** [comment]
To add to this... for a non-existent DNS server, fallback does not occur.

```sh
~> docker run --rm -p 7000:53/udp hezhijie0327/dnsproxy:latest -u tls://1.1.1.7 -f tls://8.8.8.8
2022/12/30 12:25:09 [info] Starting dnsproxy 0.46.5-ZHIJIE-DC6B0FE2
2022/12/30 12:25:09 [info] Fallback 0 is tls://8.8.8.8:853
2022/12/30 12:25:09 [info] Starting the DNS proxy server
2022/12/30 12:25:09 [info] dnsproxy: cache: disabled
2022/12/30 12:25:09 [info] Creating the UDP server socket
2022/12/30 12:25:09 [info] Listening to udp://[::]:53
2022/12/30 12:25:09 [info] Creating a TCP server socket
2022/12/30 12:25:09 [info] Listening to tcp://[::]:53
2022/12/30 12:25:09 [info] Entering the UDP listener loop on [::]:53
2022/12/30 12:25:09 [info] Entering the tcp listener loop on [::]:53
```

```sh
❱ dig @10.0.0.10 -p 7000 www.google.com

; <<>> DiG 9.18.1-1ubuntu1.2-Ubuntu <<>> @10.0.0.10 -p 7000 www.google.com
; (1 server found)
;; global options: +cmd
;; connection timed out; no servers could be reached
```

**Comment by cbuijs** [comment]
Mind that when DNSSEC validation fails, most if not all DNS servers return SERVFAIL as a valid answer (validation failed).

**Comment by chmichael** [comment]
Please fix this! If you use a custom or 3rd Party DNS server and it goes down it should query from a fallback as it should!

**Comment by abcbarryn** [comment]
I can't get fallback to do anything useful either. I shutdown the primary server so it was COMPLETELY offline and I could NOT get DNSproxy to fall back to any of the servers specified with -f. :(

**Comment by aidanharris** [comment]
I really think there should be an option somewhere to treat SERVFAIL as a network error meaning it uses the configured fallback resolver. I added `-f https://dns.adguard-dns.com/dns-query` as a fallback expecting it would be used but it never is if the upstream returns a valid SERVFAIL. This seems counter-intuitive to me even if it is technically a valid response.

**Comment by c-michail** [comment]
+1
